gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java

```diff
@@ -26,31 +27,32 @@
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public class HaGatewayManager
         implements GatewayBackendManager
 {
     private final GatewayBackendDao dao;
+    private final Supplier<List<GatewayBackend>> gatewayBackendSupplier;
     private final AtomicReference<List<GatewayBackend>> cache = new AtomicReference<>();
 
     public HaGatewayManager(Jdbi jdbi)
     {
         dao = requireNonNull(jdbi, "jdbi is null").onDemand(GatewayBackendDao.class);
-        cache.set(ImmutableList.of());
-        fetchAllBackendsToCache();
+        gatewayBackendSupplier = Suppliers.memoizeWithExpiration(dao::findAll, 500, MILLISECONDS);
```

Comment:
Should probably let the cache TTL be configurable since there is a perf vs staleness tradeoff here